src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -459,18 +459,44 @@ async Task<bool> NextResult(bool async, bool isConsuming = false, CancellationTo
                     continue;
                 }
 
-                if ((Command.WrappingBatch is not null || StatementIndex is 0) && Command.InternalBatchCommands[StatementIndex]._parameters?.HasOutputParameters == true)
+                if ((Command.WrappingBatch is not null || StatementIndex is 0) && Command.InternalBatchCommands[StatementIndex] is { HasOutputParameters: true } command)
                 {
                     // If output parameters are present and this is the first row of the resultset,
                     // we must always read it in non-sequential mode because it will be traversed twice (once
                     // here for the parameters, then as a regular row).
-                    msg = await Connector.ReadMessage(async).ConfigureAwait(false);
+                    msg = await Connector.ReadMessage(async, dataRowLoadingMode: DataRowLoadingMode.NonSequential).ConfigureAwait(false);
                     ProcessMessage(msg);
                     if (msg.Code == BackendMessageCode.DataRow)
                     {
+                        Debug.Assert(RowDescription != null);
+                        Debug.Assert(State == ReaderState.BeforeResult);
+
                         try
                         {
-                            PopulateOutputParameters(Command.InternalBatchCommands[StatementIndex]._parameters!);
+                            // Temporarily set our state to InResult and non-sequential to allow us to read the values, and in any order.
+                            var isSequential = _isSequential;
+                            var currentPosition = Buffer.ReadPosition;
+                            State = ReaderState.InResult;
+                            _isSequential = false;
```

Comment:
To support reading generically we must allow the parameters to be read out of order to support the two pass matching we do.

This is what originally led me down the road to an incorrect re-implementation.